src/Nethermind/Nethermind.AuRa/Contracts/ValidatorContract.cs

```diff
@@ -0,0 +1,142 @@
+            ///
+            /// Also called when the contract is first enabled for consensus. In this case,
+            /// the "change" finalized is the activation of the initial set.
+            /// function finalizeChange();
+            /// </summary>
+            public static readonly AbiEncodingInfo finalizeChange =
+                new AbiEncodingInfo(AbiEncodingStyle.IncludeSignature, new AbiSignature(nameof(finalizeChange)));
+
+            /// <summary>
+            /// Issue this log event to signal a desired change in validator set.
+            /// This will not lead to a change in active validator set until
+            /// finalizeChange is called.
+            ///
+            /// Only the last log event of any block can take effect.
+            /// If a signal is issued while another is being finalized it may never
+            /// take effect.
+            ///
+            /// _parent_hash here should be the parent block hash, or the
+            /// signal will not be recognized.
+            /// event InitiateChange(bytes32 indexed _parent_hash, address[] _new_set);
+            /// </summary>
+            private const string initiateChangeEventSignature = "InitiateChange(bytes32,address[])";
+            public static Keccak initiateChangeEventHash = Keccak.Compute(initiateChangeEventSignature);
+
+            /// <summary>
+            /// Get current validator set (last enacted or initial if no changes ever made)
+            /// function getValidators() constant returns (address[] _validators);
+            /// </summary>
+            public static readonly AbiEncodingInfo getValidators =
+                new AbiEncodingInfo(AbiEncodingStyle.IncludeSignature, new AbiSignature(nameof(getValidators)));
+
+            public static readonly AbiEncodingInfo addressArrayResult = new AbiEncodingInfo(AbiEncodingStyle.None,
+                new AbiSignature(nameof(addressArrayResult), new AbiArray(AbiType.Address)));
+        }
+
+        public ValidatorContract(IAbiEncoder abiEncoder, Address contractAddress) : base(contractAddress)
+        {
+            _abiEncoder = abiEncoder ?? throw new ArgumentNullException(nameof(abiEncoder));
+            _finalizeChangeTransactionData = _abiEncoder.Encode(Definition.finalizeChange);
+            _getValidatorsTransactionData = _abiEncoder.Encode(Definition.getValidators);
+        }
+
+        public Transaction FinalizeChange() => GenerateTransaction(_finalizeChangeTransactionData);
+
+        public Transaction GetValidators() => GenerateTransaction(_getValidatorsTransactionData);
+
+        public bool CheckInitiateChangeEvent(Address contractAddress, Block block, TxReceipt[] receipts, out Address[] addresses)
+        {
+            var logEntry = new LogEntry(contractAddress, 
+                Array.Empty<byte>(),
+                new[] {Definition.initiateChangeEventHash, block.ParentHash});
+
+            if (block.Bloom.IsMatch(logEntry))
```

Comment:
Maybe it would be worth to introduce this method on Block? FindLogs(...) and on a receipt ?

Reply:
Generally yes, but as block is not self-storing receipts and as there might be different log equality criteria (including data for example or not all topics), this might not be as trivial and simple as hoped.